shared/editor/nodes/Heading.ts

```diff
@@ -277,11 +278,35 @@ export default class Heading extends Node {
     return [new FoldingHeadersPlugin(this.editor.props.id), plugin];
   }
 
-  inputRules({ type }: { type: NodeType }) {
-    return this.options.levels.map((level: number) =>
-      textblockTypeInputRule(new RegExp(`^(#{1,${level}})\\s$`), type, () => ({
-        level,
-      }))
+  inputRules({ type, schema }: { type: NodeType; schema: Schema }) {
+    return this.options.levels.map(
+      (level: number) =>
+        new InputRule(
+          new RegExp(`^(#{1,${level}})\\s$`),
+          (state, _match, start, end) => {
+            const $start = state.doc.resolve(start);
+            if (
+              !$start
+                .node(-1)
+                .canReplaceWith($start.index(-1), $start.indexAfter(-1), type)
+            ) {
+              return null;
+            }
+
+            const tr = state.tr
+              .delete(start, end)
+              .setBlockType(start, start, type, { level });
+
+            const range = tr.doc.resolve(start).blockRange();
+            const wrappers =
+              range && findWrapping(range, schema.nodes["toggle_block"]);
+            if (!wrappers) {
+              return null;
+            }
+
+            return tr.wrap(range!, wrappers);
```

Comment:
Attempted this through `inputRules` on the `ToggleBlock` node and using `wrappingInputRule` but, couldn't make it work because the required regex was similar to that of `Heading` node and observed that only one of the input rules executes, depending on the order of nodes in `shared/editor/nodes/index.ts`. So, had to put the wrapping login inside of `Heading` node itself.

_Note to self_: A cleaner way could be to _detect_ when a particular node is created/inserted, then, capture that transaction and `tr.wrap` that particular node inside a toggle block. That way, the logic could be moved out of the node and can be shared across multiple nodes. Although, couldn't find a direct way to detect node creation/insertion.

Reply:
Can we leave Headings how are they are and implement toggle block entirely separately? The toggle mechanism is quite different

Reply:
Hmm, so heading won't be wrapped in a toggle block by default? Is the toggle block supposed to be triggered via command menu, for example?

Reply:
Toggle is a new element inserted through the block menu that lets you show and hide it's children. We can't change the structure of existing headings as it would modify existing docs. Also, the logic is quite different as heading toggle iterates through siblings until it finds a heading of the _same level or above_ rather than just children

Reply:
> We can't change the structure of existing headings as it would modify existing docs.

Noted 👍

> the logic is quite different as heading toggle iterates through siblings until it finds a heading of the same level or above rather than just children

Yeah, I mistakenly presumed this needed changing altogether with toggle block, which would make it easy to toggle _just_ its children. Thanks for the clarity 🙏